torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -49,8 +49,23 @@ kir::Bool* UnrollPass::getThreadPredicate(const kir::TensorView* tv) {
 }
 
 void UnrollPass::handle(kir::Expr* expr) {
-  // If tv op, predicate it (except for top level expressions)
-  if (ir_utils::isTVOp(expr) && !for_loops_.empty()) {
+  if (for_loops_.empty() && ir_utils::isTVOp(expr)) {
+    // Special handling for top level output expressions that still
+    // need predicates. One motivating example is a reduction op that
+    // reduces to a scalar (issue #491)
+    const auto out_tv = expr->outputs()[0]->as<kir::TensorView>();
+    if (out_tv->memoryType() == MemoryType::Global) {
+      const auto pred = getThreadPredicate(out_tv);
+      if (!pred->isConst() || !(pred->isConst() && pred->value().value())) {
+        kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+        kir::IfThenElse* inline_ite =
+            ir_builder.create<kir::IfThenElse>(pred, nullptr);
+        inline_ite->thenBody().push_back(expr);
+        loop_replacement_map_.insert({expr, inline_ite});
+      }
+    }
+  } else if (ir_utils::isTVOp(expr) && !for_loops_.empty()) {
+    // If tv op, predicate it (except for top level expressions)
     const auto out_tv = expr->outputs()[0]->as<kir::TensorView>();
     const auto pred = PredicateCompute::getInlinePredicate(
```

Comment:
Inline predicate today returns a True bool if there's no for loops, but probably should return the thread predicate now. I think that wasn't done before because of some other bug that Naoya fixed recently, CC @naoyam . I think it should be safe to change inline predicate. After we do that we should merge this else with your new else, then just predicate how we adding the expression with `for_loops_.empty()`:
```.cpp
if (for_loops_.empty()){
  loop_replacement_map_.insert({expr, inline_ite});
} else {
  for_loops_.back()->body().insert_before(expr, inline_ite);
  for_loops_.back()->body().erase(expr);
```

Reply:
I don't remember why it wasn't before, shouldn't the thread predicate be also used for shared memory?

Reply:
I don't know, I defer to your judgement on this one.

Reply:
I tried to see what would happen with shared memory, but I'm having different errors. First of all, we don't support allocations of 0-dim tensors (#508). I also hit another error around syncthreads insertion (#507).

For the mean time, I still think we should use thread predication both for shared and global memories, so please update the PR accordingly, @shmsong

Reply:
Refactored this part and changed `getInlinePredicate` to return thread predicate instead of `true` when at top level. Did a quick check on generated kernel in the tests, didn't seem to produce side effects other than some equivalent re-ordering of generated predicates. Also reviewed places where `getInlinePredicate` gets used and this should be safe replacement at this point.

Reply:
@naoyam Thanks for the suggestion. Added `MemoryType::Shared` in the predicate logic. #508 and #507 could be taken care of in a different PR and thanks for tracking these issues.